### DISCUSSION TITLE
Is it possible to use InnerEye fully on a local GPU machine (without Azure ML)

### DISCUSSION THREAD

**Discussion by furtheraway** [microsoft/InnerEye-DeepLearning#842] [Q&A] [answered]
Based on the [this documentation](https://innereye-deeplearning.readthedocs.io/md/building_models.html?highlight=from%20local%20checkpoints#from-local-checkpoints), model training and testing can be done on a local machine. But the example of [inference](https://innereye-deeplearning.readthedocs.io/md/building_models.html?highlight=from%20local%20checkpoints#running-a-registered-azureml-model-on-a-single-image-on-the-local-disk) is only applied to a model registered in Azure ML. 

I wonder if there is a way to do inference on the same local machine, since this will greatly reduce the efforts in securing the environment when PHI is involved.

**Comment by peterhessey** [comment, marked as answer]
Hi @furtheraway, thanks for opening this thread! The short answer to question is yes, absolutely 😄 All of InnerEye is designed to work locally identically to how it works in AzureML.

When you complete your training locally, the trained model ensemble and all the files you need for running inference will be saved to the `final_model/` folder. In there you will find a file, `score.py`. This file is the entry point for running inference on a single file that is used by the `submit_for_inference.py` script provided in the InnerEye-DeepLearning repository.

cd-ing into the `final_model/` and running the following command will give you the functionality you're looking for:

```shell
score.py --data_folder <path_to_data_folder> --image_files <data_file(s)> --use_dicom <True/False>
```

There are other optional arguments that can be used to run this script, but you've highlighted a hole in the documentation here that I will create an issue for and try to address myself over the next couple of days. You _should_ be able to view all the args with the `--help` flag but I'm not able to verify this myself right now. I'll put more details in this thread over the coming days!

**Reply by peterhessey** [reply]
Running `score.py` with the `--help` flag gives the following output:

```shell
python score.py --help
PYTHONPATH: None
usage: score.py [-h] [--data_folder DATA_FOLDER] [--model_folder MODEL_FOLDER] [--image_files IMAGE_FILES] [--result_image_name RESULT_IMAGE_NAME]
                [--use_gpu USE_GPU | --no-use_gpu] [--use_dicom [USE_DICOM]] [--result_zip_dicom_name RESULT_ZIP_DICOM_NAME] [--model_id MODEL_ID]

optional arguments:
  -h, --help            show this help message and exit
  --data_folder DATA_FOLDER
                        Path to the folder that contains the images that should be scored (default: /home/phessey/Repos/InnerEye-DeepLearning)
  --model_folder MODEL_FOLDER
                        Path to the folder that contains the model, in particular inference configuration and checkpoints. Defaults to the folder
                        where the current file lives. (default: )
  --image_files IMAGE_FILES
                        The name of the images channels to run the pipeline on. These files must exist in the data_folder. (default:
                        ['test.nii.gz'])
  --result_image_name RESULT_IMAGE_NAME
                        The name of the result image, created in the project root folder. (default: segmentation.nii.gz)
  --use_gpu USE_GPU     If GPU should be used or not. (default: True)
  --no-use_gpu
  --use_dicom [USE_DICOM]
                        If images to be scored are DICOM and output to be DICOM-RT. If this is set then image_files should contain a single zip
                        file containing a set of DICOM files. (default: False)
  --result_zip_dicom_name RESULT_ZIP_DICOM_NAME
                        The name of the zipped DICOM-RT file if use_dicom set. (default: segmentation.dcm.zip)
  --model_id MODEL_ID   The AzureML model ID. This is added to the SoftwareVersions DICOM tag in the DICOM-RT output (default: )
```
From these options you should easily be able to configure your local inference however you need! As always, happy to help with any issues / questions that arise from this.

Peter
